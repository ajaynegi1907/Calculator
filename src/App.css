:root{
  --shadow-upper: 6px 6px 16px 0 rgba(5, 5, 5, 0.438);
  --shadow-lower: -6px -6px 16px 0 rgba(235, 223, 223, 0.226);
}


body{
  background-color: rgb(51, 48, 48);
  padding: auto;
  margin: auto;
}
.App {
display: flex;
flex-direction: column;
justify-content: space-between;
border-radius: 20px;
width: 350px;
height: 550px;
margin: 40px auto;
padding: 40px 24px 32px;
box-shadow: var(--shadow-upper), var(--shadow-lower);
}

.display{
  color: white;
  width: 100%;
  height: 150px;
  overflow-y: auto;
  overflow-x: auto;
  text-align: right;
  box-shadow: var(--shadow-upper), var(--shadow-lower);
  font-weight: bold;
  font-size: 24px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
}

.keyboard {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-gap: 1.2em;
}

.keyboard .btn {
  border: none;
  background-color: rgb(51, 48, 48);
  color: black;
  cursor: grab;
  text-align: center;
  width: 60px;
  height: 60px;
  display: grid;
  place-content: center;
  border-radius: 50%;
  box-shadow: var(--shadow-upper), var(--shadow-lower);
  font-weight: bold;
  font-size: 24px;
}

.keyboard .btn-numeric{
  color: white;
}

.keyboard .btn-action{
  color: grey;
}

.keyboard .btn-operator {
  color: green
}

.keyboard .btn-calculate {
  background-color: green;
  color: black;
}

.keyboard .zero {
  grid-column: 1/3;
  width: 100%;
  border-radius: 10px;
}

.display-result {
 font-size: 16px;
 color: gray;
 padding: 8px;
}

.display-expression {
  padding: 8px;
}

.history-container {
  background-color: rgb(134, 120, 120);
  z-index: 1;
  position: absolute;
  left: inherit;
  margin: auto;
  padding: 32px 16px;
  height:  500px;
  width: 260px;
  border-radius: 20px;
  display: flex;
  flex-direction: column;
  justify-content: top;
}

.history-close-btn {
  padding: 5px;
  position: absolute;
  top: 20px;
  left: 20px;
}


.history-record{
  display: flex;
  flex-direction: column;
  text-align: right;
  padding: 5px;
  margin: 16px;
  box-shadow: var(--shadow-upper), var(--shadow-lower);
}
.history-record .expression {
  font-size: 16px;
  color: rgb(245, 228, 228);
}

.history-record .result {
  font-size: 14px;
  color: green;
}

.show-history-btn{
  border: none;
  background-color: rgb(51, 48, 48);
  color: green;
  border-radius: 16px;
  width: 60px;
  position: absolute;
  top: 0;
  margin: 52px auto;
  box-shadow: var(--shadow-upper), var(--shadow-lower);
}

.clear-history-btn {
  border: none;
  width: 100%;
  text-align: center;
  background-color: green;
  box-shadow: var(--shadow-upper), var(--shadow-lower);
  border-radius: 16px;
  font-size: 24px;
  margin: 30px auto;
}

.history-body {
  overflow-y: auto;
  overflow-x: auto;
}