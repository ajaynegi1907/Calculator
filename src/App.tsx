import "./App.css";
import Keyboard from "./components/Keyboard";
import { HistoryRecord } from "./components/History";
import { DisplayContainer } from "./containers/DisplayContainer";
import { HistoryContainer } from "./containers/History-container";
export interface AppProps{
  expression: string;
  historyData: HistoryRecord[];
  isHistoryVisible: boolean;
  setHistoryData: (data: HistoryRecord[]) => void;
  setCurrentExpression: (value: string) => void;
  setResult: (value?: number) => void;
  setHistoryPaneVisibility: (value: boolean) => void;
}

function App(props: AppProps) {
 
  const { expression, historyData} = props;
  const calculateResultClick = () => {
      try{
        const value: number = eval(expression);
        if(value !== undefined){
          if(!isFinite(value))
          alert("Invalid expression !!");
          else {
            const record: HistoryRecord = {expression: expression, result: value};
            historyData.push(record);
            props.setHistoryData(historyData);
            props.setCurrentExpression(value.toString());
            props.setResult(undefined);
          }
        }
      }
      catch{
        alert("Invalid expression !!");
      }
  }

  const clearAll = () => {
    props.setCurrentExpression("");
    props.setResult(undefined);
  }

  const removePreviousToken = () => {
    if(expression.length > 0){
      let updatedExpression = expression.slice(0, -1);
      try{
        const result: number = eval(updatedExpression);
        if(isFinite(result))
          props.setResult(result);
        else
          props.setResult(undefined);
      }
      catch{
        props.setResult(undefined);
      }
      props.setCurrentExpression(updatedExpression);
    }
  }

  const onAddTokenToExpression = (token: string) => {
    let updatedExpression = expression.concat(token);
   props.setCurrentExpression(updatedExpression);
    try{
      const result: number = eval(updatedExpression);
      if(isFinite(result))
        props.setResult(result);
    }
    catch{
      props.setResult(undefined);
    }
  }

  const toggleHistoryPaneVisibility = () => {
    props.setHistoryPaneVisibility(!props.isHistoryVisible);
  }

  return (
      <div className="App">
        <button onClick={toggleHistoryPaneVisibility} className= "show-history-btn">History </button>
        <DisplayContainer />
        <Keyboard calculateResultClick ={calculateResultClick} clearAll={clearAll} removePreviousToken={removePreviousToken} onTokenButtonClick={onAddTokenToExpression}/>
        {props.isHistoryVisible && <HistoryContainer onClose= {toggleHistoryPaneVisibility}  />}
      </div>
  );
}

export default App;
