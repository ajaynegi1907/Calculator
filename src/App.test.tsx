import App, { AppProps } from "./App";
import { shallow, ShallowWrapper } from "enzyme";

let wrapper: ShallowWrapper;
const props: AppProps = {
  expression: "",
  historyData: [],
  isHistoryVisible: false,
  setHistoryData: jest.fn(),
  setCurrentExpression: jest.fn(),
  setResult: jest.fn(),
  setHistoryPaneVisibility: jest.fn(),
};

beforeEach(() => {
  wrapper = shallow(<App {...props} />);
});

afterEach(() => {
  jest.clearAllMocks();
});

it("should match snapshot", () => {
  expect(wrapper).toMatchSnapshot();
});

it("should toggle history pane when history button is clicked", () => {
  const historyButton = wrapper.find(".show-history-btn");
  historyButton.simulate("click");
  const mockedMethod = (props.setHistoryPaneVisibility) as jest.Mock;
  expect(mockedMethod.mock.calls[0][0]).toBeTruthy();
  // wrapper.setProps({ isHistoryVisible: true });
  // historyButton.simulate("click");
  // console.log(mockedMethod.mock.calls)
  // expect(mockedMethod.mock.calls[1][0]).toBeFalsy();
});
