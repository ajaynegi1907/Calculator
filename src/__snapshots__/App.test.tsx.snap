// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="App"
>
  <button
    className="show-history-btn"
    onClick={[Function]}
  >
    History 
  </button>
  <DisplayContainer />
  <Keyboard
    calculateResultClick={[Function]}
    clearAll={[Function]}
    onTokenButtonClick={[Function]}
    removePreviousToken={[Function]}
  />
</div>
`;
